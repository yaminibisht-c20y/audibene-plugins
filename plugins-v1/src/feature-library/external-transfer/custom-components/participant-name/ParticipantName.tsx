import * as React from 'react';
import styled from 'react-emotion';
import { withTheme } from '@twilio/flex-ui';

import CallsService from '../../../../utils/serverless/Calls/CallsService';

const Name = styled('div')`
  font-size: 14px;
  font-weight: bold;
  margin-top: 10px;
  margin-bottom: 4px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
`;

const NameListItem = styled('div')`
  font-size: 12px;
  font-weight: bold;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

class ParticipantName extends React.Component<any, any> {
  state = {
    name: '',
  };

  async componentDidMount() {
    const { participant, task } = this.props;

    if (participant.participantType === 'customer') {
      this.setState({ name: task.attributes.outbound_to || task.attributes.name });
      return;
    }

    if (participant.participantType === 'unknown') {
      const response = await CallsService.getCallProperties(participant.callSid);

      if (response && response.data && response.data.call) {
        const { call } = response.data;
        const name = call.to || 'unknown';
        this.setState({ name });
      }
    } else {
      this.setState({ name: participant.worker ? participant.worker.fullName : 'unknown' });
    }
  }

  render() {
    return this.props.listMode ? (
      <NameListItem className="ParticipantCanvas-Name">{this.state.name}</NameListItem>
    ) : (
      <Name className="ParticipantCanvas-Name">{this.state.name}</Name>
    );
  }
}

export default withTheme(ParticipantName);
