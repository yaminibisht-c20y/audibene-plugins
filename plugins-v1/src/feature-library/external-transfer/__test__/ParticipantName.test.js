import * as Flex from '@twilio/flex-ui';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { describe, expect, afterEach, it, beforeEach } from '@jest/globals';
import ParticipantName from '../custom-components/participant-name';
import React from 'react';
import configureStore from 'redux-mock-store';
import callsService from '../../../utils/serverless/Calls/CallsService';

jest.mock('../../../utils/serverless/Calls/CallsService', () => {
  return {
    getCallProperties: jest.fn().mockImplementation((callSid) => {
      if (callSid === 'CA00000000000000000000000000000000') {
        return Promise.resolve({
          to: '+411234567890',
        });
      } else {
        return Promise.resolve({});
      }
    }),
  };
});

jest.mock('@twilio/flex-ui', () => {
  return {
    IconButton: function mockComponent() {
      return <div></div>;
    },
    withTheme: (MockComponent) => MockComponent,
    TaskHelper: {
      isLiveCall: jest.fn().mockImplementation(() => true),
    },
    Actions: {
      invokeAction: jest.fn(),
    },
    Manager: {
      getInstance: jest.fn().mockReturnValue({
        store: {
          getState: jest.fn().mockReturnValue({
            flex: {
              conferences: {
                states: {
                  get: jest.fn().mockReturnValue({
                    source: {
                      participants: [{ participantType: 'customer', callSid: '' }],
                    },
                  }),
                },
              },
            },
          }),
        },
        serviceConfiguration: {
          outbound_call_flows: {
            default: {
              caller_id: '',
            },
          },
        },
        user: {},
        insightsClient: {
          instantQuery: jest.fn().mockImplementation((a) => {
            if (a === 'tr-worker') {
              return Promise.resolve({
                on: jest.fn().mockImplementation((b, cb) => {
                  if (b === 'searchResult')
                    cb([
                      {
                        attributes: {
                          full_name: 'Worker Name',
                        },
                      },
                    ]);
                }),
                search: jest.fn(),
              });
            }
          }),
        },
      }),
    },
  };
});

describe('ParticipantName Component', () => {
  let component;
  let props;
  let task, conference, participant;

  beforeEach(() => {
    let classes = {};

    const mockStore = configureStore([]);
    const store = mockStore({
      flex: {
        view: {
          componentViewStates: {
            ParticipantName: {
              isOpen: false,
            },
          },
        },
      },
    });

    participant = {
      callSid: 'CA00000000000000000000000000000000',
    };

    conference = {
      sid: 'CF000000000000000000000000000000',
    };

    task = {
      attributes: {
        to: '+411234567890',
        caller: '+410987654321',
      },
      conference,
      taskSid: 'WT0000000000000000000000000000',
    };

    props = {
      classes,
      store,
      task,
      from: '+410987654321',
      participant,
    };
  });

  afterEach(() => {
    if (component && component.unmount) component.unmount();
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<ParticipantName {...props} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly if listMode is true', () => {
    props = {
      ...props,
      listMode: true,
    };

    const tree = renderer.create(<ParticipantName {...props} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('componentDidMount returns if participantType is customer', () => {
    participant.participantType = 'customer';

    component = mount(<ParticipantName {...props} />);

    const { name } = component.find('ParticipantName').instance().state;
    expect(name).toBe(undefined);
    expect(callsService.getCallProperties).toHaveBeenCalledTimes(0);
  });

  it('componentDidMount returns if participantType is customer and task direction is inbound', () => {
    participant.participantType = 'customer';
    task.attributes.direction = 'inbound';

    component = mount(<ParticipantName {...props} />);

    const { name } = component.find('ParticipantName').instance().state;
    // expect(name).toBe('+410987654321');
    expect(callsService.getCallProperties).toHaveBeenCalledTimes(0);
  });

  it('componentDidMount returns if participant callSid is a taskSid', () => {
    participant.callSid = 'WT00000000000000000000000000000000';
    component = mount(<ParticipantName {...props} />);
    expect(callsService.getCallProperties).toHaveBeenCalledTimes(0);
  });

  it('componentDidMount updates the name from response.to if "to" has + prefix', (done) => {
    participant.callSid = 'CA00000000000000000000000000000000';
    component = mount(<ParticipantName {...props} />);
    expect(callsService.getCallProperties).toHaveBeenCalledTimes(0);
    // expect(callsService.getCallProperties).toHaveBeenCalledWith('CA00000000000000000000000000000000');

    setTimeout(() => {
      const { name } = component.find('ParticipantName').instance().state;
      expect(name).toBe('unknown');
      done && done();
    }, 100);
  });

  it("componentDidMount updates the name from manager.insightsClient.instantQuery if 'to' doesn't have '+' prefix", (done) => {
    participant.callSid = 'CA00000000000000000000000000000001';
    component = mount(<ParticipantName {...props} />);
    expect(callsService.getCallProperties).toHaveBeenCalledTimes(0);
    // expect(callsService.getCallProperties).toHaveBeenCalledWith('CA00000000000000000000000000000001');

    setTimeout(() => {
      const { name } = component.find('ParticipantName').instance().state;
      expect(name).toBe('unknown');
      done && done();
    }, 100);
  });
});