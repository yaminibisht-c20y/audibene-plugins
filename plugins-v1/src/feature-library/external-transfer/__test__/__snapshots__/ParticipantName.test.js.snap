// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantName Component renders correctly 1`] = `
<div
  className="ParticipantCanvas-Name css-2sxfea"
>
  unknown
</div>
`;

exports[`ParticipantName Component renders correctly if listMode is true 1`] = `
<div
  className="ParticipantCanvas-Name css-1a0arli"
>
  unknown
</div>
`;
